core/src/connection/pool.rs

```diff
@@ -581,40 +600,22 @@ where
                                             .map_or(0, |conns| conns.len());
                         if let Err(e) = self.limits.check_established(current) {
                             let connected = entry.close();
-                            return Poll::Ready(PoolEvent::PendingConnectionError {
+                            let num_established = e.current;
+                            return Poll::Ready(PoolEvent::ConnectionError {
                                 id,
-                                endpoint: connected.endpoint,
-                                peer: Some(connected.info.peer_id().clone()),
-                                error: PendingConnectionError::ConnectionLimit(e),
+                                connected,
+                                error: ConnectionError::ConnectionLimit(e),
+                                num_established,
                                 pool: self,
-                                handler: None,
                             })
                         }
-                        // Check peer ID.
-                        if let Some(peer) = peer {
-                            if &peer != entry.connected().peer_id() {
-                                let connected = entry.close();
-                                return Poll::Ready(PoolEvent::PendingConnectionError {
-                                    id,
-                                    endpoint: connected.endpoint,
-                                    peer: Some(connected.info.peer_id().clone()),
-                                    error: PendingConnectionError::InvalidPeerId,
-                                    pool: self,
-                                    handler: None,
-                                })
+                        // Peer ID checks must already have happened. See `add_pending`.
+                        if cfg!(debug_assertions) {
+                            assert_ne!(&self.local_id, entry.connected().peer_id());
+                            if let Some(peer) = peer {
+                                assert_eq!(&peer, entry.connected().peer_id());
```

Comment:
I suppose that this is because you require `Debug` above?
If so, I'd prefer to not enforce `Debug` if possible.

```suggestion
                            assert_ne!(&self.local_id, entry.connected().peer_id(), "Unexpected local peer ID");
                            if let Some(peer) = peer {
                                assert_eq!(&peer, entry.connected().peer_id(), "PeerId mismatch");
```

Reply:
`fmt::Debug` is required by `assert_eq!` and `assert_ne!` for the arguments, so are you suggesting something like:
```rust
if &self.local_id == entry.connected().peer_id() {
    panic!("...")
}
if let Some(peer) = peer {
    if peer != entry.connected().peer_id() {
        panic!("...")
    }
}
```
? It is unfortunate though not to have the problematic peer IDs in the output. Why is it undesirable to require `fmt::Debug`?

Reply:
I have now done this: https://github.com/libp2p/rust-libp2p/pull/1493/commits/5807155992cba55e18a1cd5e41d2b1b9b68e6163

Reply:
> Why is it undesirable to require fmt::Debug?

I just see it as "more correct" to only enforce the minimum possible set of traits.
In an ideal world, we would print the `PeerId` if it implements `Debug` and not print it if it doesn't. But that's unfortunately not possible.